### ISSUE TITLE
Unexpected compiler error when building std for custom Windows target

### ISSUE THREAD

**Issue by MarnixKuijs** [rust-lang/rust#81228]
When building the standard library for a custom Windows target I get unexpected compiler errors. The errors appear when compiling  the `compiler_builtins` crate. When looking into `compiler_builtins` there is [a cfg](https://github.com/rust-lang/compiler-builtins/blob/63ccaf11f08fb5d0b39cc33884c5a1a63f547ace/src/probestack.rs#L46) to prevent the file at fault from compiling on Windows platforms. As far as my knowledge goes this cfg directive should work as long as the `target_family` is Windows, which is set to inside of the target json. When my target is set to `x86_64-pc-windows-msvc` everything compiles fine.

### Meta
```
Rustc: 1.49.0
Compiling: stable branch
```

<details><summary>Backtrace</summary>
<p>

```
error: unknown directive
            .pushsection .text.__rust_probestack
            ^
error: expected absolute expression
            .type  __rust_probestack, @function
                   ^
error: unknown directive
            .hidden __rust_probestack
            ^
error: unknown directive
            .size __rust_probestack, . - __rust_probestack
            ^
error: unknown directive
            .popsection
            ^
error: unknown directive
            .pushsection .text.__rust_probestack
            ^
error: expected absolute expression
            .type  __rust_probestack, @function
                   ^
error: unknown directive
            .hidden __rust_probestack
            ^
error: unknown directive
            .size __rust_probestack, . - __rust_probestack
            ^
error: unknown directive
            .popsection
            ^
error: unknown directive
  |
note: instantiated into assembly here
 --> <inline asm>:2:13
  |
2 |             .pushsection .text.__rust_probestack
  |             ^

error: expected absolute expression
  |
note: instantiated into assembly here
 --> <inline asm>:4:20
  |
4 |             .type  __rust_probestack, @function
  |                    ^

error: unknown directive
  |
note: instantiated into assembly here
 --> <inline asm>:5:13
  |
5 |             .hidden __rust_probestack
  |             ^

error: unknown directive
   |
note: instantiated into assembly here
  --> <inline asm>:57:13
   |
57 |             .size __rust_probestack, . - __rust_probestack
   |             ^

error: unknown directive
   |
note: instantiated into assembly here
  --> <inline asm>:58:13
   |
58 |             .popsection
   |             ^

error: aborting due to 5 previous errors

error: could not compile `compiler_builtins`
```

</p>
</details>

**Comment by MarnixKuijs** [comment]
Upon further investigation it isn't a bug in the building process. The problem lies in the target spec json. I based my target spec on the one you get from running `rustc +nightly -Z unstable-options --print target-spec-json`, which in my case give the spec from `x86_64-pc-windows-msvc`, which I than started modifying. In this spec there is the following entry: `target-family = "windows"`. This however results in the check `#![cfg(windows)]` to fail. To make it pass the entry should be `target_family = "windows". I can imagine other people can get stuck on this.

We can either close this issue and open a new one  or repurpose this issue to discuss the problems that `target-family` vs `target_family` have in the target spec.
